crates/volta-migrate/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+//! Provides types for modeling the current state of the Volta directory and for migrating between versions
+//!
+//! A new layout should be represented by it's own struct (as in the existing v0 or v1 modules)
+//! Migrations between types should be represented by `TryFrom` implementations between the layout types
+//! (see v1.rs for examples)
+//!
+//! NOTE: Since the layout file is written once the migration is complete, all migration implementations
+//! need to be aware that they may be partially applied (if something fails in the process) and should be
+//! able to re-start gracefully from an interrupted migration
+
+#![cfg(feature = "volta-updates")]
+use std::collections::HashSet;
+use std::convert::TryInto;
+use std::fs::{DirEntry, Metadata};
+use std::path::Path;
+
+mod empty;
+mod v0;
+mod v1;
+
+use v0::V0;
+use v1::V1;
+
+use log::debug;
+use volta_core::error::ErrorDetails;
+use volta_core::fs::read_dir_eager;
+use volta_core::layout::volta_home;
+#[cfg(unix)]
+use volta_core::layout::volta_install;
+use volta_core::shim;
+use volta_fail::{Fallible, ResultExt};
+
+enum MigrationState {
+    Empty(empty::Empty),
+    V0(Box<V0>),
+    V1(Box<V1>),
+}
+
+impl MigrationState {
+    fn current() -> Fallible<Self> {
+        /*
+        Triage for determining the current layout version:
+        - Does the V1 layout file exist? If yes then V1
+        - Does Volta Home exist? If yes (Windows) then V0
+            - If yes (Unix) then check if Volta Install is outside shim_dir?
+                - If yes, then V0
+                - If no, then check if $VOLTA_HOME/load.sh exists? If yes then V0
+        - Else Empty
+
+        The extra logic on Unix is necessary because Unix installs can be either inside or outside $VOLTA_HOME
+        If it is inside, then the directory necessarily must exist, so we can't use that as a determination.
+        If it is outside (and for Windows which is always outside), then if $VOLTA_HOME exists, it must be from a
+        previous, V0 installation.
+
+        Going forward, each new version will have an associated layout file, so we can use that as a discriminant
```

Comment:
Given the point made in this comment ("going forward, ..."), it feels like it would be more readable to split this top-level algorithm into two parts that delegate to two separate helper functions:

- infer legacy state
- detect branded state

The more complex logic is inferring the legacy state, so it's nice to cordon that off from the simple part.

Reply:
I like that separation a lot, I'll add that!